### ISSUE TITLE
Help

### ISSUE THREAD

**Issue by mhmmd10** [Dhyeythumar/video-streamer#4]
Hi Bro
i want to run this
but iam beginner
how to set video
i need help bro.

**Comment by dhyeythumar** [comment]
Hi @mhmmd10,
Have you tried running the notebook given in this repo?
This one: https://github.com/dhyeythumar/mlagents-video-streamer/blob/v2.0/Streaming%20ML-Agents%20from%20Colab%20-v2.0.ipynb

**Comment by mhmmd10** [comment]
yes 
it is works
And it gets stream
But how to use video in it?

**Comment by dhyeythumar** [comment]
Are you talking about storing video locally?
Because I didn't understood that did you meant by asking `But how to use video in it?`

**Comment by mhmmd10** [comment]
I want to use video for live streaming

**Comment by dhyeythumar** [comment]
Video is already lived streamed when you run the notebook, you also mention the your favorite server URL & secret where you would like to stream it.

**Comment by mhmmd10** [comment]
for example
I want to add game.mp4 video from path Google Drive.
To play this video live stream.
Is it possible to do this?
What should be done?
HOW?

**Comment by mhmmd10** [comment]
Did you understand?

**Comment by dhyeythumar** [comment]
Yes, I got your point.
So you want to do something like upload a video on Google Drive then connect Drive to Google Colab & mention the video path & then live stream this video to any server, right?

**Comment by mhmmd10** [comment]
yes

**Comment by dhyeythumar** [comment]
Actually, I created this package to live stream the ML-Agents training process from Google Colab to any server. And for this I am capturing the frames of the training process & streaming it to the mentioned server.

TL;DR
The functionality which you are looking for is not provided by this package.

But I think this can be done & might be a new project idea which I might consider to develop.

**Comment by mhmmd10** [comment]
OKay

Thanks

**Comment by dhyeythumar** [comment]
I am closing this issue. If you want any further assistance then feel free to reopen it or create a new one.

**Comment by mhmmd10** [comment]
Hello, do you have a quick solution so I can do this?

**Comment by dhyeythumar** [comment]
I think what you can do is:

1. Connect Drive to Colab
2. Read the video file frame by frame
3. Now use this streaming code (you might have to make changes to it)
The below code is from my [gym-video-streamer](https://github.com/dhyeythumar/gym-video-streamer) repo

```
class Streamer(object):
    def __init__(self, frame_shape, frames_per_sec, output_frames_per_sec, streamURL):
        self.proc = None
        # Frame shape should be lines-first, so w and h are swapped
        h, w, pixfmt = frame_shape
        if pixfmt != 3 and pixfmt != 4:
            raise error.InvalidFrame(
                "Your frame has shape {}, but we require (w,h,3) or (w,h,4), i.e., \
                RGB values for a w-by-h image, with an optional alpha channel.".format(frame_shape))


        self.wh = (w, h)
        self.includes_alpha = (pixfmt == 4)
        self.frame_shape = frame_shape
        self.frames_per_sec = frames_per_sec
        self.output_frames_per_sec = output_frames_per_sec
        self.ffmpegOutput = ""


        # this it will stream
        if streamURL != "":
            self.ffmpegOutput = streamURL
        # if not possible then record the video & save it locally
        else:
            directory = "./videos"
            if not os.path.exists(directory):
                print("\nCreating directory for storing videos {}".format(directory))
                os.makedirs(directory, exist_ok=True)


            self.__directory = os.path.abspath(directory)
            date = datetime.now().strftime("(%Y.%m.%d)-(%I.%M.%S)-%p")
            self.ffmpegOutput = os.path.join(self.__directory, 'video_{}.mp4'.format(date))
            touch(self.ffmpegOutput)  # just in case


        if shutil.which("ffmpeg") is not None:
            self.backend = "ffmpeg"
        else:
            raise error.DependencyNotInstalled("No ffmpeg executable found!")
        self.start()


    def start(self):
        self.cmdline = [self.backend,
                        "-nostats",
                        "-loglevel", "error",  # suppress warnings
                        '-y',
                        "-threads:v", "4",
                        "-filter_threads", "4",


                        # input
                        "-f", "rawvideo",
                        "-s:v", "{}x{}".format(*self.wh),  # size of one frame
                        "-pix_fmt", ("rgb32" if self.includes_alpha else "rgb24"),
                        "-framerate", "%d" % self.frames_per_sec,
                        "-i", "-",  # The input comes from a pipe


                        "-b:v", "600k",  # bitrate
                        # "-minrate:v", "600k",
                        # "-maxrate:v", "600k",
                        # "-bufsize:v", "600k",


                        # output
                        "-vcodec", "libx264",
                        "-pix_fmt", "yuv420p",
                        "-r:v", "%d" % self.output_frames_per_sec,


                        "-crf", "18",  # lower CRF values correspond to higher bitrates,
                        "-bf:v", "3",  # set maximum number of B frames between non-B-frames
                        "-refs:v", "16",  # reference frames to consider for motion compensation
                        "-f", "flv",
                        self.ffmpegOutput
                        ]
        self.proc = subprocess.Popen(self.cmdline, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)


    def capture_frame(self, frame):
        if not isinstance(frame, (np.ndarray, np.generic)):
            raise error.InvalidFrame('Wrong type {} for {} (must be np.ndarray or np.generic)'.format(type(frame), frame))
        if frame.shape != self.frame_shape:
            raise error.InvalidFrame("Your frame has shape {}, but the VideoRecorder is configured for shape {}.".format(frame.shape, self.frame_shape))
        if frame.dtype != np.uint8:
            raise error.InvalidFrame("Your frame has data type {}, but we require uint8 (i.e. RGB values from 0-255).".format(frame.dtype))


        # write this frame to the writable stream of subprocess
        out = self.proc.stdin.write(frame.tobytes())
        if out is None:
            raise error.Error("ffmpeg streaming broken & gave '{}' error".format(out))


    def close(self):
        self.proc.stdin.close()
        ret = self.proc.wait()
        if ret != 0:
            printMsg("Streamer encoder exited with status {}".format(ret))
```

Video complete code here: https://github.com/dhyeythumar/gym-video-streamer/blob/v1.0/gym_video_streamer/streamer.py#L21-L109

**Comment by mhmmd10** [comment]
oh
can't run this :(
